### PR TITLE
Allow table/column comments to be specified in the model

### LINKED ISSUE #15037: Allow table/column comments to be specified in the model
env:
    EntityFrameworkCore.Tools 2.2.3
	MySql.Data.EntityFrameworkCore 8.0.15

``` SQL

DROP TABLE IF EXISTS `demo`;
CREATE TABLE `demo`  (
  `aaa` varchar(150) CHARACTER SET utf8 COLLATE utf8_general_ci NOT NULL COMMENT 'HOW TO ADD COMMENT',
  --                                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  `bbb` varchar(32) CHARACTER SET utf8 COLLATE utf8_general_ci NOT NULL,
  PRIMARY KEY (`aaa`) USING BTREE
) ENGINE = InnoDB CHARACTER SET = utf8 COLLATE = utf8_general_ci ROW_FORMAT = Compact;

````

### PR DESCRIPTION
**Pull request by Muppets**
Summary of the changes
 - Added HasComment extensions to table and column
 - Added SQL server support for comments in extended properties

Fixes #15037

### CODE REVIEW

File: src/EFCore.Relational/Extensions/RelationalPropertyBuilderExtensions.cs [hunk 1]
```diff
@@ -442,5 +442,74 @@ public static class RelationalPropertyBuilderExtensions
...
                 fromDataAnnotation);
+
+        /// <summary>
+        ///     Configures a comment to be applied to the column
+        /// </summary>
+        /// <param name="propertyBuilder"> The builder for the property being configured. </param>
+        /// <param name="comment"> The comment for the column. </param>
+        /// <returns> The same builder instance so that multiple calls can be chained. </returns>
+        public static PropertyBuilder HasComment(
+            [NotNull] this PropertyBuilder propertyBuilder,
```
Comment (bricelam) [inline src/EFCore.Relational/Extensions/RelationalPropertyBuilderExtensions.cs]: IMHO, these are specific to SQL Server. I don't think we should add them to every relational provider.

**Comment by bricelam** [comment]
We've also had a request to add these by-convention based on `DescriptionAttribute`... we should discuss with the team in a design meeting.

File: src/EFCore.Relational/Extensions/RelationalPropertyBuilderExtensions.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Extensions/RelationalPropertyBuilderExtensions.cs]: Seems that they are supported on PostgreSQL, Oracle an MySQL

File: src/EFCore.Relational/Extensions/RelationalEntityTypeBuilderExtensions.cs [hunk 2]
```diff
@@ -462,5 +462,47 @@ public static class RelationalEntityTypeBuilderExtensions
...
+            {
+                return null;
+            }
+
+            entityTypeBuilder.Metadata.SetComment(comment, fromDataAnnotation);
+            return entityTypeBuilder;
+        }
+
+        /// <summary>
+        ///     Returns a value indicating whether the schema of the view or table name can be set for this entity type
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Extensions/RelationalEntityTypeBuilderExtensions.cs]: Fix docs

File: src/EFCore.Relational/Extensions/RelationalEntityTypeExtensions.cs [hunk 3]
```diff
@@ -253,5 +253,40 @@ public static void SetSchema([NotNull] this IMutableEntityType entityType, [CanB
...
+
+        /// <summary>
+        ///     Returns the comment for the column this property is mapped to.
+        /// </summary>
+        /// <param name="entityType"> The entity type. </param>
+        /// <returns> The comment for the column this property is mapped to. </returns>
+        public static string GetComment([NotNull] this IEntityType entityType)
+        {
+            var value = (string)entityType[RelationalAnnotationNames.Comment];
+            if (value != null)
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Extensions/RelationalEntityTypeExtensions.cs]: Redundant check

**Comment by AndriySvyryd** [comment]
You also need to add a check to `RelationalModelValidator.ValidateSharedTableCompatibility` and `RelationalModelValidator.ValidateSharedColumnsCompatibility`

File: src/EFCore.Relational/Extensions/RelationalPropertyBuilderExtensions.cs (hunk 1 above)
Comment (bricelam) [reply src/EFCore.Relational/Extensions/RelationalPropertyBuilderExtensions.cs]: Neat. I though Oracle might. Didn't expect so many other to too.

**Comment by Muppets** [comment]
@AndriySvyryd @bricelam This should be good to go. Let me know if I've missed anything else.

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs [hunk 4]
```diff
@@ -249,6 +249,19 @@ protected virtual void ValidateDefaultValuesOnKeys([NotNull] IModel model, [NotN
                                 otherKey.Properties.Format()));
                     }
 
+                    var currentComment = entityType.GetComment() ?? "";
+                    var previousComment = nextEntityType.GetComment() ?? "";
+                    if (!currentComment.Equals(previousComment, StringComparison.OrdinalIgnoreCase))
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs]: These should use `StringComparison.Ordinal` for consistency and some minor perf gain.

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs [hunk 5]
```diff
@@ -249,6 +249,19 @@ protected virtual void ValidateDefaultValuesOnKeys([NotNull] IModel model, [NotN
                                 otherKey.Properties.Format()));
                     }
 
+                    var currentComment = entityType.GetComment() ?? "";
+                    var previousComment = nextEntityType.GetComment() ?? "";
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs]: We can skip validation on `null` comments, just store the first non-null one and compare every other one against it.

File: src/EFCore.SqlServer/Migrations/SqlServerMigrationsSqlGenerator.cs [hunk 6]
```diff
@@ -1616,6 +1628,123 @@ protected override void ForeignKeyAction(ReferentialAction referentialAction, Mi
...
+            Check.NotNull(builder, nameof(builder));
+            Check.NotNull(table, nameof(table));
+
+            if (comment != null)
+            {
+                EndStatement(builder);
+
+                GenerateCreateExtendedProperty(builder, model, "Comment", comment, schema, table, columnName);
+            }
+
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Migrations/SqlServerMigrationsSqlGenerator.cs]: Extra line

File: src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs (hunk 5 above)
Comment (Muppets) [reply src/EFCore.Relational/Infrastructure/RelationalModelValidator.cs]: Like this?

**Comment by AndriySvyryd** [comment]
Thanks for your contribution!

**Pull request merged**